archeryutils/handicaps/handicap_scheme_aa.py

```diff
@@ -0,0 +1,296 @@
+
+Extended Summary
+----------------
+Code to calculate information using a number of schemes:
+
+- Original Archery Australia (J Park)
+- Updated Archery Australia (J Park)
+
+Routine Listings
+----------------
+- HandicapAA Class
+- HandicapAA2 Class
+
+References
+----------
+- Modelling archers’ scores at different distances to quantify
+  score loss due to equipment selection and technique errors
+  Park, J (2014)
+  https://doi.org/10.1177%2F1754337114539308
+
+"""
+
+import numpy as np
+
+from .handicap_scheme import FloatArray, HandicapScheme
+
+
+class HandicapAA(HandicapScheme):
+    """
+    Class to represent the original Archery Australia scheme by J. Park.
+
+    Parameters
+    ----------
+    ang_0 : float
+        datum angle used in scheme, default = 1.0e-3,
+    k0 : float
+        handicap offset to set scratch point, default= 2.37,
+    ks : float
+        change with each handicap step as a geometric progression, default= 0.027,
+    kd : float
+        distance scaling factor, default= 0.004,
+
+    Attributes
+    ----------
+    name : str="AA"
+        The name of the handicap scheme.
+    arw_d_out : float
+        diameter of an outdoor arrow: 5.0e-3 [metres]
+    arw_d_in: float
+        diameter of an indoor arrow 9.3e-3 [metres]
+
+    Methods
+    -------
+    sigma_t(handicap, dist)
+        Calculate angular deviation for given handicap and distance.
+    sigma_r(handicap, dist)
+        Calculate radial deviation for a given handicap and distance.
+    arrow_score(target, handicap, arw_d=None)
+        Calculate the average arrow score for a given target and handicap.
+    score_for_passes(rnd, handicap, arw_d=None, rounded_score=True)
+        Calculate the expected score for all passes in a round for a given handicap.
+    score_for_round(rnd, handicap, arw_d=None, rounded_score=True)
+        Calculate the expected score for a round for a given handicap.
+    handicap_from_score(score,rnd, arw_d=None, int_prec=False)
+        Calculate the handicap for a given score on a given round.
+
+    Warnings
+    --------
+    Using non-default values for the kwargs may produce results that
+    do not match the official values for this scheme.
+    Only change if you are experimenting and know what you are doing!
+
+    See Also
+    --------
+    HandicapScheme : The base class for a handicap scheme from which this is subclassed.
+
+    """
+
+    def __init__(
+        self,
+        ang_0: float = 1.0e-3,
+        k0: float = 2.37,
+        ks: float = 0.027,
+        kd: float = 0.004,
+    ):
+        self.params = {
+            "ang_0": ang_0,  # Baseline angle used for group size 1.0 [millirad].
+            "k0": k0,  # Offset required to set handicap 100 at desired score.
+            "ks": ks,  # Change with each step of geometric progression.
+            "kd": kd,  # Distance scaling factor [1/metres].
+        }
+
+        self.arw_d_out = 5.0e-3
+
+        self.name = "AA"
```

Comment:
I don't really like how `name`, `arw_d_out`, and `desc_scale` are class variables in the super class but instance variances in the subclass. It means `HandicapAA.name == "unnamed"`, but `HandicapAA().name == "AA"`. I'd recommend either keeping these as class variables or as instance variables throughout. To have them as instance variables, you could set them in the super class `__init__` function, then start every subclass `__init__` with `super().__init__()`.

Just overriding the class variables in the subclasses would work too, or if you wanted to get fancy and make them (pseudo-)immutable you could implement them as properties. 

You could also get really weird and set the class variables using [`__init_subclass__`](https://www.geeksforgeeks.org/__init_subclass__-in-python/) and set them in subclasses like:

 ```python
class HandicapAA(HandicapScheme, name="AA", arw_d_out=5.0e-3, desc_scale=False):
    ...
```
I think it'd look something like the following in the base class:

```python

def __init_subclass__(cls, name: str, arw_d_out: float=5.5e-3, arw_d_in: float=9.3e-3, desc_scale: bool=True):
    super().__init_subclass__() # not sure if this is needed?
    cls.name = name # must be set in subclasses
    cls.arw_d_in = arw_d_in # optional
    cls.arw_d_out = arw_d_out # optional
    cls.desc_scale = desc_scale # optional
```

It might be overkill, but the advantage is similar to that of `@abstractmethod` in that it forces subclasses to set a name. The main disadvantage is that it's a fairly obscure bit of Python and might just cause developer confusion further down the road.

Reply:
Nice spot, thanks! - I have been working on this late at nights so missed that!
They should indeed be set as instance variables rather than class variables (I think is my preference), so I'll put an `__init__` in `HandicapScheme` and then call the super from the subclass `__init__`.

Reply:
https://github.com/jatkinson1000/archeryutils/pull/78/commits/8e5440f7af3a87edd8bb4a49901bd483bedca504 should fix this I think.

Reply:
@LiamPattinson Is this resolved to your satisfaction?

Reply:
This is resolved, and in-fact somewhat moot as all instance variables are now explicitly set in each subclass so I will mark as resolved.